pkg/streamingpromql/planning.go

```diff
@@ -440,6 +477,51 @@ func findFunction(name string) (functions.Function, bool) {
 	return functions.FUNCTION_UNKNOWN, false
 }
 
+// functionNeedsDeduplication checks if a function needs deduplication and merge operator.
+// This is determined by whether the function drops the __name__ label or otherwise manupulates it.
+func functionNeedsDeduplication(fnc functions.Function) bool {
+	switch fnc {
+	case
+		// Time transformation functions
+		functions.FUNCTION_DAY_OF_MONTH, functions.FUNCTION_DAY_OF_WEEK,
+		functions.FUNCTION_DAY_OF_YEAR, functions.FUNCTION_DAYS_IN_MONTH,
+		functions.FUNCTION_HOUR, functions.FUNCTION_MINUTE, functions.FUNCTION_MONTH,
+		functions.FUNCTION_YEAR,
+		// Range vector functions
+		functions.FUNCTION_AVG_OVER_TIME, functions.FUNCTION_CHANGES,
+		functions.FUNCTION_COUNT_OVER_TIME, functions.FUNCTION_DELTA,
+		functions.FUNCTION_DERIV, functions.FUNCTION_IDELTA, functions.FUNCTION_INCREASE,
+		functions.FUNCTION_IRATE,
+		functions.FUNCTION_MAX_OVER_TIME, functions.FUNCTION_MIN_OVER_TIME,
+		functions.FUNCTION_PRESENT_OVER_TIME, functions.FUNCTION_QUANTILE_OVER_TIME,
+		functions.FUNCTION_RATE, functions.FUNCTION_RESETS, functions.FUNCTION_STDDEV_OVER_TIME,
+		functions.FUNCTION_STDVAR_OVER_TIME, functions.FUNCTION_SUM_OVER_TIME,
+		// Instant vector transformations
+		functions.FUNCTION_ABS, functions.FUNCTION_ACOS, functions.FUNCTION_ACOSH,
+		functions.FUNCTION_ASIN, functions.FUNCTION_ASINH, functions.FUNCTION_ATAN,
+		functions.FUNCTION_ATANH, functions.FUNCTION_CEIL, functions.FUNCTION_COS,
+		functions.FUNCTION_COSH, functions.FUNCTION_DEG, functions.FUNCTION_EXP,
+		functions.FUNCTION_FLOOR, functions.FUNCTION_LN, functions.FUNCTION_LOG10,
+		functions.FUNCTION_LOG2, functions.FUNCTION_RAD, functions.FUNCTION_SGN,
+		functions.FUNCTION_SIN, functions.FUNCTION_SINH, functions.FUNCTION_SQRT,
+		functions.FUNCTION_TAN, functions.FUNCTION_TANH,
+		functions.FUNCTION_CLAMP, functions.FUNCTION_CLAMP_MAX, functions.FUNCTION_CLAMP_MIN,
+		functions.FUNCTION_ROUND, functions.FUNCTION_PREDICT_LINEAR, functions.FUNCTION_TIMESTAMP,
+		functions.FUNCTION_DOUBLE_EXPONENTIAL_SMOOTHING,
+		// Histogram functions
+		functions.FUNCTION_HISTOGRAM_AVG,
+		functions.FUNCTION_HISTOGRAM_COUNT, functions.FUNCTION_HISTOGRAM_FRACTION,
+		functions.FUNCTION_HISTOGRAM_QUANTILE,
+		functions.FUNCTION_HISTOGRAM_STDDEV, functions.FUNCTION_HISTOGRAM_STDVAR,
+		functions.FUNCTION_HISTOGRAM_SUM,
+		// Label manipulation functions
+		functions.FUNCTION_LABEL_JOIN, functions.FUNCTION_LABEL_REPLACE:
```

Comment:
[nit] I would format each of these on their own line, rather than having some on their own line and some with multiple on the same line